audio/common/wasapi.c

```diff
@@ -0,0 +1,740 @@
+      case AUDCLNT_E_BUFFER_ERROR:
+         return "AUDCLNT_E_BUFFER_ERROR";
+      case AUDCLNT_E_BUFFER_OPERATION_PENDING:
+         return "AUDCLNT_E_BUFFER_OPERATION_PENDING";
+      case AUDCLNT_E_BUFFER_SIZE_NOT_ALIGNED:
+         return "AUDCLNT_E_BUFFER_SIZE_NOT_ALIGNED";
+      case AUDCLNT_E_BUFFER_SIZE_ERROR:
+         return "AUDCLNT_E_BUFFER_SIZE_ERROR";
+      case AUDCLNT_E_CPUUSAGE_EXCEEDED:
+         return "AUDCLNT_E_CPUUSAGE_EXCEEDED";
+      case AUDCLNT_E_DEVICE_IN_USE:
+         return "AUDCLNT_E_DEVICE_IN_USE";
+      case AUDCLNT_E_DEVICE_INVALIDATED:
+         return "AUDCLNT_E_DEVICE_INVALIDATED";
+      case AUDCLNT_E_ENDPOINT_CREATE_FAILED:
+         return "AUDCLNT_E_ENDPOINT_CREATE_FAILED";
+      case AUDCLNT_E_EXCLUSIVE_MODE_NOT_ALLOWED:
+         return "AUDCLNT_E_EXCLUSIVE_MODE_NOT_ALLOWED";
+      case AUDCLNT_E_INVALID_DEVICE_PERIOD:
+         return "AUDCLNT_E_INVALID_DEVICE_PERIOD";
+      case AUDCLNT_E_INVALID_SIZE:
+         return "AUDCLNT_E_INVALID_SIZE";
+      case AUDCLNT_E_NOT_INITIALIZED:
+         return "AUDCLNT_E_NOT_INITIALIZED";
+      case AUDCLNT_E_OUT_OF_ORDER:
+         return "AUDCLNT_E_OUT_OF_ORDER";
+      case AUDCLNT_E_SERVICE_NOT_RUNNING:
+         return "AUDCLNT_E_SERVICE_NOT_RUNNING";
+      case AUDCLNT_E_UNSUPPORTED_FORMAT:
+         return "AUDCLNT_E_UNSUPPORTED_FORMAT";
+      case AUDCLNT_E_WRONG_ENDPOINT_TYPE:
+         return "AUDCLNT_E_WRONG_ENDPOINT_TYPE";
+      /* AUDCLNT success codes */
+      case AUDCLNT_S_BUFFER_EMPTY:
+         return "AUDCLNT_S_BUFFER_EMPTY";
+      /* Something else; probably from an API that we started using
+       * after mic support was implemented */
+      default:
+         return "<unknown>";
+   }
+}
+
+const char *wave_subtype_name(const GUID *guid)
+{
+   if (IsEqualGUID(guid, &KSDATAFORMAT_SUBTYPE_IEEE_FLOAT))
+   {
+      return "KSDATAFORMAT_SUBTYPE_IEEE_FLOAT";
+   }
+
+   return "<unknown sub-format>";
+}
+
+const char *wave_format_name(const WAVEFORMATEXTENSIBLE *format)
+{
+   switch (format->Format.wFormatTag)
+   {
+      case WAVE_FORMAT_PCM:
+         return "WAVE_FORMAT_PCM";
+      case WAVE_FORMAT_EXTENSIBLE:
+         return wave_subtype_name(&format->SubFormat);
+      default:
+         return "<unknown>";
+   }
+}
+
+const char* wasapi_error(DWORD error)
+{
+   static char error_message[256];
+
+   FormatMessage(
+         FORMAT_MESSAGE_IGNORE_INSERTS |
+         FORMAT_MESSAGE_FROM_SYSTEM,
+         NULL,
+         error,
+         MAKELANGID(LANG_ENGLISH, SUBLANG_DEFAULT),
+         error_message,
+         sizeof(error_message) - 1,
+         NULL);
+
+   return error_message;
+}
+
+static const char* wasapi_data_flow_name(EDataFlow data_flow)
+{
+   switch (data_flow)
+   {
+      case eCapture:
+         return "eCapture";
+      case eRender:
+         return "eRender";
+      case eAll:
+         return "eAll";
+      default:
+         return "<unknown>";
+   }
```

Comment:
There should always be a return at the end of a function with a return type. I'd suggest turning the default case into a break; and then returning "<unknown>" at the end of the function.